### ISSUE TITLE
Exception "Request failed: The server committed a protocol violation. Section=ResponseStatusLine" with multi-step availability test

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/aspnetcore#28613]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/Exception-Request-failed:-The-server-co/1278145)._

---
Hi,

we have a problem with a multi-step availability test both on Visual Studio 2019 Enterprise and Azure Application Insights.

The first HTTP request works as expected, but the second HTTP request ends with the following exception:

Request failed: The server committed a protocol violation. Section=ResponseStatusLine

I have done the following tests trying to find a solution to this issue:

1) Tested the same HTTP requests flow with Postman and they works fine;
2) Tested the same HTTP requests flow with cURL and they works fine;
3) Tested the same HTTP requests flow with Visual Studio 2019 Enterprise as a webtest project and got the same exception at the second HTTP request call. If i use Fiddler or Burp in the middle of the HTTP calls, I don't get any exception and also the Visual Studio 2019 webtest project run works fine. I managed also (without any proxy in behind) to temporarily use a TLS cipher that permits Wireshark decryption with the private key of the HTTPS server SSL certificate (by temporarily disabling all ciphers that supports the Diffie-Hellman key exchange) and checked the trace of the HTTP requests, but I haven't find any useful information, the HTTP flow seems correct at first glance (HTTP headers are corrects , and also JSON data returned and HTTP status).

Have you any suggestion on how I can better investigate the reason behind this exception?

Just to add another information, we are currently using Dynatrace Synthetic with the same HTTP flow and it works fine, and we are trying to configure Azure Application Insights with the same services in order to evaluate if we can switch from Dynatrace to Azure Application Insights.

Thanks in advance for all your replies!

---
#### Feedback Bot on 12/6/2020, 06:23 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 12/6/2020, 06:52 PM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

**Comment by BrennanConroy** [comment]
Can you share the wireshark trace with us?

**Comment by Tratcher** [comment]
"The server committed a protocol violation. Section=ResponseStatusLine" is an error reported by HttpWebRequest from .NET 4.x. The fact that the response works when passed through a proxy implies there's a formatting incompatibility between the server and client and the proxies are fixing it up. The wireshark trace will show us any formatting oddities.

**Comment by valeriodelsarto** [comment]
Here you can find the decrypted wireshark trace, I had to remove from the trace the frame 18 (which included a login password within an HTTP header), for this reason also the second HTTP request that was previously decrypted as frame 23 is now encrypted as frame 22, probably there are some packet missing due to the fact that I've removed the frame 18. However, the interesting part in the trace is the frame 29 (previously was frame 30), which includes the HTTP response that cause the "The server committed a protocol violation. Section=ResponseStatusLine" Exception on both Visual Studio webtest and Azure Application Insights. The first HTTP request (the "login" phase) works fine within Visual Studio, is the second request that ends with this Exception. As you can see in frame 30 and 31, the client close the TCP session when this Exception is logged client side parsing the HTTP response.

Link to the wireshark trace: https://progensoftitsrl-my.sharepoint.com/:u:/g/personal/vdelsarto_apparound_it/EVGjE32EHSlJknbkfUkq78oBB_VJEzteGcUYTk1S-b4cBg

Link password: Y_ukbglqri4Vo3RIPw1E

**Comment by Tratcher** [comment]
I think the problem is actually in frame 20. The response has a stated content-length of 337, but after the 337 body bytes there's a stray 0x0A (Line Feed). How did that get there? The server committed a protocol violation by sending more data than it said it was going to. The client only reads the expected amount of data and saves then moves onto the next request. The remainder is parsed as the start of the next response and triggers that ResponseStatusLine parsing error. Some clients have a mitigation against this kind of violation where they fully drain the connection if they haven't sent the next request yet. That would explain why using a proxy mitigates the issue.

How are you generating the response in frame 20? Can you show the code that writes it?

The headers say this is an IIS server and that you're using openresty?

@jkotalik doesn't our IIS in-proc code enforce that the content-length matches the response body size?

**Comment by valeriodelsarto** [comment]
Hi, you're right, I didn't noticed that there was a Line Feed char at the end of the http data object in the response of the first http request (the "login" API in order to get a valid token), but I didn't totally understood how it could impact the Exception that I get when the code is parsing the response of the second http request (the "request for some data" API, sending a valid token). I mean that this Exception is raised when Visual Studio is going to parse the http response that is present in Frame 29. Let me premise that I'm more a System Administrator rather than a Developer, so I'm not so expert of .NET code and libraries, but I managed to find the same Line Feed char at the end of the first http request in all the traces that I have done using a proxy in the middle, with both Fiddler and Burp, and in both cases Visual Studio didn't raised this kind of Exception.

If needed for further investigation, I can ask to my colleagues if I can share part of the code that generates this response, meanwhile what I can say is that for my understanding the application that is generating the second http response is an ASP.NET application wrote in C# that runs on IIS 10 on Windows Server 2016, with an Nginx reverse proxy in front of the application that runs within the OpenResty platform.

**Comment by Tratcher** [comment]
> but I didn't totally understood how it could impact the Exception that I get when the code is parsing the response of the second http request

1. The client reads the headers for the first response. Those headers say the body is 337 bytes long.
1. The client reads 337 bytes for the body. It does not read the 338th byte, that stays in the buffer.
1. The next request is sent and the client starts reading the next response. That response has a leftover LF at the front of it because it was never consumed as part of the prior request, making it fail to parse. KaBoom.

Some proxies and clients mitigate this because after step 2 they drain any leftover data off the connection and discard it rather than forwarding it to the client or treating it as part of the next response. That means step 3 starts with a clean connection. Another mitigation some proxies and clients have is to be tolerant of extra CRs and LFs at the start of a response and a proxy would re-format the response without those.

Note this is a protocol issue more than a .NET issue, your app is committing a protocol violation by sending too much data and some clients will handle that better than others. HttpWebRequest in this example is quite strict.

**Comment by valeriodelsarto** [comment]
Thanks, understood. I'll try to explain it to my colleagues that developed that part in order to get rid of this additional LF byte at the end of the 1st request, I didn't think they could be treated in the same buffer and then cause this kind of Exception.

**Comment by valeriodelsarto** [comment]
We solved this issue as you suggested, we were wrongly using a ngx.say instead of a ngx.print inside a LUA script in our Nginx reverse proxy, and this was causing the additional LF char to be sent back to the client.

Now the Visual Studio webtest is working fine and I'm quite confident that Azure Application Insights will do the same.

**Comment by Tratcher** [comment]
Oh, so the response was correctly generated by IIS but it was the proxy that broke it?

For triage: We should still at least add a test for proper body lengths in IIS.

**Comment by valeriodelsarto** [comment]
Right, there was a couple of redirects within the reverse proxy in front of the application that built the final response to the user by doing some checks and requesting the token to a .NET application on IIS, this step creates a correct response but the reverse proxy was returning an additional LF char due to the wrong method used within a LUA script.
